File: Second Year/COS212/Assignments/07_01_Assignment 4/Assignment 1_JAVA_FILES/PriorityQueue.java
public class PriorityQueue<T extends Comparable<T>> extends OrderedContainer<T> {

    public PriorityQueue(LinearStructure<T> c) {
        super(c);
    }

    public PriorityQueue(PriorityQueue<T> other) {
        super(other.dataStructure);
    }

    @Override
    public PriorityQueue<T> clone() {
        return new PriorityQueue<T>(this.dataStructure);
    }

    public T remove() {
        /*int pos = 0;
        T priority = null;
        try {
            priority = next();
        } catch (RemoveException ignored) {

        }
        while (true) {
            try {
                T temp = this.dataStructure.remove(pos);
                if (temp == priority) {
                    break;
                }
                else this.dataStructure.insert(pos, temp);
                pos++;
            }
            catch (RemoveException e) {
                break;
            }
        }
        return priority;*/

        T temp = null;
        try {
            temp = this.getImplementation().remove(0);
        } catch (RemoveException ignored) {

        }
        return temp;
    }

    public T next() throws RemoveException {
        /*int pos = 0;
        T highest = null;
        try {
            highest = this.dataStructure.remove(0);
            this.dataStructure.insert(0, highest);
            while (true) {
                try {
                    T temp = this.dataStructure.remove(pos);
                    this.dataStructure.insert(pos, temp);
                    if (temp.compareTo(highest) > 0)
                        highest = temp;
                    pos++;
                }
                catch (RemoveException e) {
                    break;
                }
            }
            return highest;
        }
        catch (RemoveException ignored) {

        }
        return highest;*/

        T temp = null;
        try {
            temp = this.getImplementation().remove(0);
            this.getImplementation().insert(0, temp);
        }
        catch (RemoveException ignored) {

        }
        return temp;
    }

    public void insert(T el) {
        int i = 0;
        while (true) {
            try {
                T temp = this.getImplementation().remove(i);
                this.getImplementation().insert(i, temp);
                if (temp.compareTo(el) < 0) {
                    break;
                } else {
                    i++;
                }
            }
            catch (RemoveException e) {
                break;
            }
        }
        try {
            this.getImplementation().insert(i, el);
        } catch (RemoveException e) {
            //e.printStackTrace();
        }
    }
}

File: First Year/COS110/Assignments/Assignment 4/priorityQueue.cpp
#include "priorityQueue.h"

template <class T>
PriorityQueue<T>::PriorityQueue(LinearStructure<T> *c) : Queue<T>(c) {
    this->front = 0;
    this->rear = 0;
    this->ascend = true;
}

template <class T>
PriorityQueue<T>::PriorityQueue(const PriorityQueue<T> &other) : Queue<T>(other.dataStructure) {
    this->front = other.front;
    this->rear = other.rear;
}

template <class T>
PriorityQueue<T> & PriorityQueue<T>::operator=(const PriorityQueue<T> &other) {
    PriorityQueue<T> * newQueue = new PriorityQueue<T>(this->dataStructure);
    return *newQueue;
}

template <class T>
PriorityQueue<T>::~PriorityQueue<T>() {

}

template <class T>
void PriorityQueue<T>::insert(T el) {
    if (!this->isEmpty()) {
        if (ascend) {
            if (this->rear == 1) {
                if (el > this->dataStructure->get(0)) {
                    this->dataStructure->insert(1, el);
                    this->rear++;
                }
                else {
                    this->dataStructure->insert(0, el);
                    this->rear++;
                }
            }
            else {
                for (int i = 0; i < this->rear; i++) {
                    //cout  << "VAL = " << i << endl;
                    if (el < this->dataStructure->get(i)) {
                        //cout << "RAN";
                        this->dataStructure->insert(i, el);
                        this->rear++;
                        return;
                    }
                }
                //cout << "HERE" << endl;
                this->dataStructure->insert(this->rear, el);
                this->rear++;
            }
        }
        else {

            if (this->rear == 1) {
                if (el < this->dataStructure->get(0)) {
                    this->dataStructure->insert(1, el);
                    this->rear++;
                }
                else {
                    this->dataStructure->insert(0, el);
                    this->rear++;
                }
            }
            else {
                if (this->rear == 1) {
                    if (el > this->dataStructure->get(0)) {
                        this->dataStructure->insert(1, el);
                        this->rear++;
                    }
                    else {
                        this->dataStructure->insert(0, el);
                        this->rear++;
                    }
                }
                else {
                    for (int i = 0; i < this->rear; i++) {
                        //cout  << "VAL = " << i << endl;
                        if (el > this->dataStructure->get(i)) {
                            //cout << "RAN";
                            this->dataStructure->insert(i, el);
                            this->rear++;
                            return;
                        }
                    }
                    //cout << "HERE" << endl;
                    this->dataStructure->insert(this->rear, el);
                    this->rear++;
                }
            }
        }
    }
    else {
        this->dataStructure->insert(0, el);
        this->rear++;
    }
}

template <class T>
void PriorityQueue<T>::reverse() {

    T * temp = new T[this->rear + 1];
    int tRear = this->rear;

    int counter = 0;
    while (!this->isEmpty()) {
        temp[counter] = this->remove();
        counter++;
    }

    Queue<T> * newQueue = new Queue<T>(this->getImplementation());
    while (!newQueue->isEmpty()) {
        newQueue->remove();
    }
    for (int i = tRear - 1; i >= 0; i--) {
        newQueue->insert(temp[i]);
    }
    this->dataStructure = newQueue->getImplementation();
    this->rear = tRear;

    this->ascend = !this->ascend;
}